source/extensions/filters/network/mysql_proxy/mysql_terminal_filter.cc

```diff
@@ -0,0 +1,243 @@
+    }
+    parent_.upstream_conn_data_ = nullptr;
+  }
+}
+
+void MySQLTerminalFilter::onPoolReady(Envoy::Tcp::ConnectionPool::ConnectionDataPtr&& conn,
+                                      Upstream::HostDescriptionConstSharedPtr) {
+  canceler_ = nullptr;
+  upstream_conn_data_ = std::move(conn);
+  upstream_event_handler_ = std::make_unique<UpstreamEventHandler>(*this);
+  upstream_conn_data_->addUpstreamCallbacks(*upstream_event_handler_);
+
+  read_callbacks_->continueReading();
+}
+
+void MySQLTerminalFilter::onPoolFailure(Tcp::ConnectionPool::PoolFailureReason reason,
+                                        Upstream::HostDescriptionConstSharedPtr host) {
+  config_->stats_.login_failures_.inc();
+
+  std::string host_info;
+  if (host != nullptr) {
+    host_info = " remote host address " + host->address()->asString();
+  }
+  switch (reason) {
+  case Tcp::ConnectionPool::PoolFailureReason::Overflow:
+    ENVOY_LOG(info, "mysql proxy upstream connection pool: too many connections, {}", host_info);
+    break;
+  case Tcp::ConnectionPool::PoolFailureReason::LocalConnectionFailure:
+    ENVOY_LOG(info, "mysql proxy upstream connection pool:onocal connection failure, {}",
+              host_info);
+    break;
+  case Tcp::ConnectionPool::PoolFailureReason::RemoteConnectionFailure:
+    ENVOY_LOG(info, "mysql proxy upstream connection pool: remote connection failure, {}",
+              host_info);
+    break;
+  case Tcp::ConnectionPool::PoolFailureReason::Timeout:
+    ENVOY_LOG(info, "mysql proxy upstream connection pool: connection failure due to time out, {}",
+              host_info);
+    break;
+  default:
+    ENVOY_LOG(info, "mysql proxy upstream connection pool: unknown error, {}", host_info);
+    break;
+  }
+  canceler_ = nullptr;
+  read_callbacks_->connection().close(Network::ConnectionCloseType::NoFlush);
+}
+
+MySQLTerminalFilter::UpstreamEventHandler::UpstreamEventHandler(MySQLTerminalFilter& filter)
+    : parent_(filter) {}
+
+void MySQLTerminalFilter::onProtocolError() {
+  MySQLMoniterFilter::onProtocolError();
+  ENVOY_LOG(info, "communication failure due to protocol error");
+  read_callbacks_->connection().close(Network::ConnectionCloseType::NoFlush);
+}
+
+void MySQLTerminalFilter::onNewMessage(MySQLSession::State state) {
+  MySQLMoniterFilter::onNewMessage(state);
+  // close connection when received message on state NotHandled.
+  if (state == MySQLSession::State::NotHandled || state == MySQLSession::State::Error) {
+    ENVOY_LOG(info, "connection closed due to unexpected state occurs on communication");
+    read_callbacks_->connection().close(Network::ConnectionCloseType::NoFlush);
+  }
+}
+
+void MySQLTerminalFilter::onServerGreeting(ServerGreeting& greet) {
+  MySQLMoniterFilter::onServerGreeting(greet);
+  ENVOY_LOG(debug, "server {} send challenge", greet.getVersion());
+  sendLocal(greet);
+}
+
+void MySQLTerminalFilter::onClientLogin(ClientLogin& client_login) {
```

Comment:
I thought that intention behind new `AuthInfo` message https://github.com/envoyproxy/envoy/blob/fb1efe4872e42ea8baa0489e986e0484eea4d192/api/envoy/extensions/filters/network/mysql_proxy/v3/mysql_proxy.proto#L22-L28 is to handle clients' authentication requests locally, but here I see that you still send the to remote.